solver/src/main/java/org/chocosolver/solver/search/strategy/selectors/variables/AbstractCriterionBasedVariableSelector.java

```diff
@@ -90,46 +58,25 @@ public Element(int count, int w0, int w1) {
      * Needed to save operations
      */
     final IEnvironment environment;
-    /**
-     * The number of conflicts which have occurred since the beginning of the search.
-     */
-    int conflicts = 0;
-    /**
-     * Stores for each propagator, its {@link Element}.
-     */
-    final HashMap<Propagator<?>, Element> failCount = new HashMap<>();
-    /**
-     * Required to store for each variable the number of watchers (ie, propagator and futvars)
-     */
-    private final HashMap<Variable, Integer> observed = new HashMap<>();
     /**
      * Scoring for each variable, is updated dynamically.
      */
-    final TObjectDoubleMap<Variable> weights = new TObjectDoubleHashMap<>(15, 1.5f, 0.);
+    final Weightable<Variable> weights;
+
     /**
-     * Refined weights,
-     * for DOMWDEG_REFINED and CHS
+     * Create a variable selector based on a criterion
+     * @param vars decision variables
+     * @param seed seed for breaking ties randomly, if set to -1, ties are broken by the variable position
+     * @param flush flush threshold, when reached, it flushes scores
      */
-    final HashMap<Propagator<?>, double[]> refinedWeights = new HashMap<>();
-    static final double[] rw = {0.};
-
-    final BiConsumer<Variable, Propagator<?>> update = (v, p) -> {
-        Element elt = failCount.get(p);
-        if (elt != null) {
-            if (p.getVar(elt.ws[0]) == v) {
-                updateFutvars(p, elt, 0);
-            } else if (p.getVar(elt.ws[1]) == v) {
-                updateFutvars(p, elt, 1);
-            }
-        }
-    };
-
     public AbstractCriterionBasedVariableSelector(V[] vars, long seed, int flush) {
         this.random = new java.util.Random(seed);
         this.solver = vars[0].getModel().getSolver();
         this.environment = vars[0].getModel().getEnvironment();
         this.last = environment.makeInt(vars.length - 1);
         this.flushThs = flush;
+//        this.weights = new Weightable.Map<>();
+        this.weights = new Weightable.Array<>();
```

Comment:
Shouldn't this depend on the maximum of the id of variables ? The Array version of Weightable seems to be a source of high memory consumption if the ids of variables are not continuous from 1 (if variables have id from 500 to 2500, `weights` will rely on an array of size 2501, but indexes 0 to 499 will never be used, which seems to be "a lot" of memory for nothing). That said, if Weightable are used only in this selector, it might not be that important

Reply:
You're suggesting to add an offset as the minimum ID among all variables to reduce the size of the array.
That's a good idea but the way AbstractCriterionBasedSelector is used allows adding weight to variables not declared initially in the scope of the instance (because for example, when a propagator fails, dom/wdeg does not ignore variable out of the declared scope).
Your remark is correct but I feel like it's going to overcomplicate a structure which goal is to be simple and efficient.
I should probably consider a way for users to switch from `Array` to `Map`.

Reply:
I was not suggesting to have an offset for the `Array` implementation of `Weightable` (as it might be very difficult, or prone to inefficiency, due to the fact that variables in scope are dynamically concidered) but merely to have a condition that chooses between the `Array` and the `Map` implementation of `Weightable` depending on the ids of vars (even knowing the variables in scope of weights are dynamic). The idea being to be more memory-efficient. But, once again, if Weightable is used only in selectors, it might not be worth the trouble (as the restricted use will not lead to a lot of memory inefficiency)

Reply:
In fact, using an Array is about 30% faster than using a Map.
This is mainly due to the access of an element in an array of doubles which is faster than retrieving the position of the key in an hashmap.
That said, that's why I suggest to Split `Model.nextId()` into `Model.nextVarId()` and `Model.nextPropId()`, to also (mainly) limit the memory footprint.